File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/mail/RightsTest.scala
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 * http://www.apache.org/licenses/LICENSE-2.0                   *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 * ***************************************************************/

package org.apache.james.jmap.mail

import org.apache.james.jmap.mail.Right.DeleteMailbox
import org.apache.james.mailbox.model.MailboxACL.{EntryKey, Rfc4314Rights => JavaRfc4314Rights, Right => JavaRight}
import org.apache.james.mailbox.model.{MailboxACL => JavaMailboxACL}
import org.scalatest.matchers.must.Matchers
import org.scalatest.wordspec.AnyWordSpec

import scala.jdk.CollectionConverters._

class RightsTest extends AnyWordSpec with Matchers {
  val NEGATIVE = true
  val USER: String = "user"
  val USER_ENTRYKEY: EntryKey = EntryKey.createUserEntryKey(USER)
  val OTHER_USER_ENTRYKEY: EntryKey = EntryKey.createUserEntryKey("otherUser")

  "Right ofCharacter" should  {
    "recognise 'a'" in {
      Right.forChar('a') must be(Some(Right.Administer))
    }
    "recognise 'e'" in {
      Right.forChar('e') must be(Some(Right.Expunge))
    }
    "recognise 'i'" in {
      Right.forChar('i') must be(Some(Right.Insert))
    }
    "recognise 'l'" in {
      Right.forChar('l') must be(Some(Right.Lookup))
    }
    "recognise 'r'" in {
      Right.forChar('r') must be(Some(Right.Read))
    }
    "recognise 'w'" in {
      Right.forChar('w') must be(Some(Right.Write))
    }
    "recognise 't'" in {
      Right.forChar('t') must be(Some(Right.DeleteMessages))
    }
    "recognise 'p'" in {
      Right.forChar('p') must be(Some(Right.Post))
    }
    "return empty when unknown" in {
      Right.forChar('k') must be(None)
    }
  }
  "From ACL" should  {
    "filter out group entries" in {
      val acl = new JavaMailboxACL(Map(
        EntryKey.createGroupEntryKey("group") -> JavaRfc4314Rights.fromSerializedRfc4314Rights("aet")).asJava)

      Rights.fromACL(MailboxACL.fromJava(acl)) must be(Rights.EMPTY)
    }
    "filter out negative users" in {
      val acl = new JavaMailboxACL(Map(
        EntryKey.createUserEntryKey(USER, NEGATIVE) -> JavaRfc4314Rights.fromSerializedRfc4314Rights("aet")).asJava)

      Rights.fromACL(MailboxACL.fromJava(acl)) must be(Rights.EMPTY)
    }
    "accept users" in {
      val acl = new JavaMailboxACL(Map(
        USER_ENTRYKEY -> JavaRfc4314Rights.fromSerializedRfc4314Rights("aet")).asJava)

      Rights.fromACL(MailboxACL.fromJava(acl)) must be(Rights.of(USER_ENTRYKEY, Seq(Right.Administer, Right.Expunge, Right.DeleteMessages)))
    }
    "filter out unknown rights" in {
      val acl = new JavaMailboxACL(Map(
        USER_ENTRYKEY -> JavaRfc4314Rights.fromSerializedRfc4314Rights("aetxk")).asJava)

      Rights.fromACL(MailboxACL.fromJava(acl)) must be(Rights.of(USER_ENTRYKEY, Seq(Right.Administer, Right.Expunge, Right.DeleteMessages, Right.DeleteMailbox)))
    }
  }
  "To ACL" should  {
    "return empty when empty" in {
      Rights.EMPTY.toMailboxAcl.asJava must be(new JavaMailboxACL())
    }
    "return acl conversion" in {
      val user1 = EntryKey.createUserEntryKey("user1")
      val user2 = EntryKey.createUserEntryKey("user2")
      val expected = new JavaMailboxACL(Map(
          user1 -> new JavaRfc4314Rights(JavaRight.Administer, JavaRight.DeleteMessages),
          user2 -> new JavaRfc4314Rights(JavaRight.PerformExpunge, JavaRight.Lookup))
        .asJava)
      val jmapPojo = Rights.of(user1, Seq(Right.Administer, Right.DeleteMessages))
        .append(user2, Seq(Right.Expunge, Right.Lookup))

      jmapPojo.toMailboxAcl.asJava must be(expected)
    }
  }
  "Remove entries" should  {
    "return empty when empty" in {
      Rights.EMPTY.removeEntriesFor(USER_ENTRYKEY) must be(Rights.EMPTY)
    }
    "return empty when only user" in {
      Rights.of(USER_ENTRYKEY, Right.Lookup)
        .removeEntriesFor(USER_ENTRYKEY) must be(Rights.EMPTY)
    }
    "only remove specified users" in {
      val expected = Rights.of(OTHER_USER_ENTRYKEY, Right.Lookup)

      Rights.of(USER_ENTRYKEY, Right.Lookup)
        .append(OTHER_USER_ENTRYKEY, Right.Lookup)
        .removeEntriesFor(USER_ENTRYKEY) must be(expected)
    }
  }
  "mayAddItems" should  {
    "return empty when no user" in {
      Rights.EMPTY.mayAddItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return false when no insert right" in {
      Rights.of(USER_ENTRYKEY, Seq(Right.Administer, Right.Expunge, Right.Lookup, Right.DeleteMessages, Right.Read, Right.Seen, Right.Write))
        .mayAddItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return true when insert right" in {
      Rights.of(USER_ENTRYKEY, Right.Insert)
        .mayAddItems(USER_ENTRYKEY) must be(Applicable)
    }
  }
  "mayReadItems" should  {
    "return empty when no user" in {
      Rights.EMPTY.mayReadItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return false when no read right" in {
      Rights.of(USER_ENTRYKEY, Seq(Right.Administer, Right.Expunge, Right.Lookup, Right.DeleteMessages, Right.Administer, Right.Seen, Right.Write))
        .mayReadItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return true when read right" in {
      Rights.of(USER_ENTRYKEY, Right.Read)
        .mayReadItems(USER_ENTRYKEY) must be(Applicable)
    }
  }
  "mayRemoveItems" should  {
    "return empty when no user" in {
      Rights.EMPTY.mayRemoveItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return false when no delete right" in {
      Rights.of(USER_ENTRYKEY, Seq(Right.Administer, Right.Expunge, Right.Lookup, Right.Read, Right.Administer, Right.Seen, Right.Write))
        .mayRemoveItems(USER_ENTRYKEY) must be(NotApplicable)
    }
    "return true when delete right" in {
      Rights.of(USER_ENTRYKEY, Right.DeleteMessages)
        .mayRemoveItems(USER_ENTRYKEY) must be(Applicable)
    }
  }
  "mayRename" should  {
    "return unsupported" in {
      Rights.EMPTY.mayRename(USER_ENTRYKEY) must be(Unsupported)
    }
  }
  "mayDelete" should  {
    "return unsupported" in {
      Rights.EMPTY.mayDelete(USER_ENTRYKEY) must be(Unsupported)
    }
  }
  "mayCreateChild" should  {
    "return unsupported" in {
      Rights.EMPTY.mayCreateChild(USER_ENTRYKEY) must be(Unsupported)
    }
  }
}

File: server/protocols/jmap-rfc-8621/src/test/java/org/apache/james/jmap/model/mailbox/RightsTest.java
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 *   http://www.apache.org/licenses/LICENSE-2.0                 *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 ****************************************************************/

package org.apache.james.jmap.model.mailbox;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.apache.james.core.Username;
import org.apache.james.jmap.model.mailbox.Rights.Right;
import org.apache.james.mailbox.model.MailboxACL;
import org.apache.james.mailbox.model.MailboxACL.Entry;
import org.apache.james.mailbox.model.MailboxACL.EntryKey;
import org.apache.james.mailbox.model.MailboxACL.Rfc4314Rights;
import org.junit.Test;

import com.google.common.collect.ImmutableMap;

import nl.jqno.equalsverifier.EqualsVerifier;

public class RightsTest {

    public static final boolean NEGATIVE = true;
    public static final Username USERNAME = Username.of("user");
    public static final Username OTHER_USERNAME = Username.of("otherUser");

    @Test
    public void rightsShouldMatchBeanContract() {
        EqualsVerifier.forClass(Rights.class)
            .verify();
    }

    @Test
    public void forCharShouldReturnRightWhenA() {
        assertThat(Right.forChar('a'))
            .isEqualTo(Right.Administer);
    }

    @Test
    public void forCharShouldReturnRightWhenE() {
        assertThat(Right.forChar('e'))
            .isEqualTo(Right.Expunge);
    }

    @Test
    public void forCharShouldReturnRightWhenI() {
        assertThat(Right.forChar('i'))
            .isEqualTo(Right.Insert);
    }

    @Test
    public void forCharShouldReturnRightWhenL() {
        assertThat(Right.forChar('l'))
            .isEqualTo(Right.Lookup);
    }

    @Test
    public void forCharShouldReturnRightWhenR() {
        assertThat(Right.forChar('r'))
            .isEqualTo(Right.Read);
    }

    @Test
    public void forCharShouldReturnRightWhenW() {
        assertThat(Right.forChar('w'))
            .isEqualTo(Right.Write);
    }

    @Test
    public void forCharShouldReturnRightWhenT() {
        assertThat(Right.forChar('t'))
            .isEqualTo(Right.DeleteMessages);
    }

    @Test
    public void forCharShouldThrowOnUnsupportedRight() {
        assertThatThrownBy(() -> Right.forChar('k'))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void fromACLShouldFilterOutGroups() throws Exception {
        MailboxACL acl = new MailboxACL(ImmutableMap.of(
            EntryKey.createGroupEntryKey("group"), Rfc4314Rights.fromSerializedRfc4314Rights("aet")));

        assertThat(Rights.fromACL(acl))
            .isEqualTo(Rights.EMPTY);
    }

    @Test
    public void fromACLShouldFilterNegatedUsers() throws Exception {
        MailboxACL acl = new MailboxACL(ImmutableMap.of(
            EntryKey.createUserEntryKey(USERNAME, NEGATIVE), Rfc4314Rights.fromSerializedRfc4314Rights("aet")));

        assertThat(Rights.fromACL(acl))
            .isEqualTo(Rights.EMPTY);
    }

    @Test
    public void fromACLShouldAcceptUsers() throws Exception {
        MailboxACL acl = new MailboxACL(ImmutableMap.of(
            EntryKey.createUserEntryKey(USERNAME), Rfc4314Rights.fromSerializedRfc4314Rights("aet")));

        assertThat(Rights.fromACL(acl))
            .isEqualTo(Rights.builder()
                .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.DeleteMessages)
                .build());
    }

    @Test
    public void fromACLShouldFilterOutUnknownRights() throws Exception {
        MailboxACL acl = new MailboxACL(ImmutableMap.of(
            EntryKey.createUserEntryKey(USERNAME), Rfc4314Rights.fromSerializedRfc4314Rights("aetpk")));

        assertThat(Rights.fromACL(acl))
            .isEqualTo(Rights.builder()
                .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.DeleteMessages)
                .build());
    }

    @Test
    public void toMailboxAclShouldReturnEmptyAclWhenEmpty() {
        Rights rights = Rights.EMPTY;

        assertThat(rights.toMailboxAcl())
            .isEqualTo(new MailboxACL());
    }

    @Test
    public void toMailboxAclShouldReturnAclConversion() throws Exception {
        String user1 = "user1";
        String user2 = "user2";
        Rights rights = Rights.builder()
            .delegateTo(Username.of(user1), Right.Administer, Right.DeleteMessages)
            .delegateTo(Username.of(user2), Right.Expunge, Right.Lookup)
            .build();

        assertThat(rights.toMailboxAcl())
            .isEqualTo(new MailboxACL(
                new Entry(user1, MailboxACL.Right.Administer, MailboxACL.Right.DeleteMessages),
                new Entry(user2, MailboxACL.Right.PerformExpunge, MailboxACL.Right.Lookup)));
    }

    @Test
    public void removeEntriesForShouldNotModifyEmptyRights() {
        assertThat(Rights.EMPTY.removeEntriesFor(USERNAME))
            .isEqualTo(Rights.EMPTY);
    }

    @Test
    public void removeEntriesForShouldRemoveUsernameEntryWhenPresent() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Lookup)
            .build();
        assertThat(rights.removeEntriesFor(USERNAME))
            .isEqualTo(Rights.EMPTY);
    }

    @Test
    public void removeEntriesForShouldOnlyRemoveSpecifiedUsername() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Lookup)
            .delegateTo(OTHER_USERNAME, Right.Lookup)
            .build();

        Rights expected = Rights.builder()
            .delegateTo(OTHER_USERNAME, Right.Lookup)
            .build();

        assertThat(rights.removeEntriesFor(USERNAME))
            .isEqualTo(expected);
    }

    @Test
    public void mayAddItemsShouldReturnEmptyWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayAddItems(USERNAME))
            .isEmpty();
    }

    @Test
    public void mayCreateChildShouldReturnUnsupportedWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayCreateChild(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }

    @Test
    public void mayDeleteShouldReturnUnsupportedWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayDelete(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }

    @Test
    public void mayReadItemsShouldReturnEmptyWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayReadItems(USERNAME))
            .isEmpty();
    }

    @Test
    public void mayRemoveItemsShouldReturnEmptyWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayRemoveItems(USERNAME))
            .isEmpty();
    }

    @Test
    public void mayRenameShouldReturnEmptyWhenNoUserRights() {
        assertThat(Rights.EMPTY.mayRename(USERNAME))
            .isEmpty();
    }

    @Test
    public void mayAddItemsShouldReturnFalseWhenNoInsertRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup,
                Right.DeleteMessages, Right.Read, Right.Seen, Right.Write)
            .build();

        assertThat(rights.mayAddItems(USERNAME))
            .contains(false);
    }

    @Test
    public void mayCreateChildShouldReturnUnsupportedWhenFullRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup, Right.Insert,
                Right.DeleteMessages, Right.Read, Right.Seen, Right.Write)
            .build();

        assertThat(rights.mayCreateChild(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }

    @Test
    public void mayDeleteShouldReturnUnsupportedWhenFullRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup, Right.Insert,
                Right.DeleteMessages, Right.Read, Right.Seen, Right.Write)
            .build();

        assertThat(rights.mayDelete(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }

    @Test
    public void mayReadItemsShouldReturnFalseWhenNoReadRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup, Right.Insert,
                Right.DeleteMessages, Right.Seen, Right.Write)
            .build();

        assertThat(rights.mayReadItems(USERNAME))
            .contains(false);
    }

    @Test
    public void mayRemoveItemsShouldReturnFalseWhenNoDeleteMessagesRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup, Right.Insert,
                Right.Read, Right.Seen, Right.Write)
            .build();

        assertThat(rights.mayRemoveItems(USERNAME))
            .contains(false);
    }

    @Test
    public void mayRenameShouldReturnFalseWhenNoWriteRights() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Administer, Right.Expunge, Right.Lookup, Right.Insert,
                Right.DeleteMessages, Right.Read, Right.Seen)
            .build();

        assertThat(rights.mayRename(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }

    @Test
    public void mayAddItemsShouldReturnTrueWhenInsertRight() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Insert)
            .build();

        assertThat(rights.mayAddItems(USERNAME))
            .contains(true);
    }

    @Test
    public void mayReadItemsShouldReturnTrueWhenReadRight() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Read)
            .build();

        assertThat(rights.mayReadItems(USERNAME))
            .contains(true);
    }

    @Test
    public void mayRemoveItemsShouldReturnTrueWhenDeleteMessagesRight() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.DeleteMessages)
            .build();

        assertThat(rights.mayRemoveItems(USERNAME))
            .contains(true);
    }

    @Test
    public void mayRenameShouldReturnTrueWhenWriteRight() {
        Rights rights = Rights.builder()
            .delegateTo(USERNAME, Right.Write)
            .build();

        assertThat(rights.mayRename(USERNAME))
            .isEqualTo(Rights.UNSUPPORTED);
    }
}